pinot-core/src/main/java/org/apache/pinot/core/query/optimizer/statement/JsonStatementOptimizer.java

```diff
@@ -387,10 +387,27 @@ private static void optimizeJsonPredicate(Expression expression, @Nullable Table
    *  @return A string array containing all the parts of an identifier. An identifier may have one or more parts that
    *  are joined together using <DOT>. For example the identifier "testTable.jsonColumn.name.first" consists up of
    *  "testTable" (name of table), "jsonColumn" (name of column), "name" (json path), and "first" (json path). The last
-   *  two parts when joined together (name.first) represent a JSON path expression.
+   *  two parts when joined together (name.first) represent a JSON path expression. If an identifier refers to a top-
+   *  level JSON array element (as in column_name[1]), then there is do <DOT> in the identifier.
    */
   private static String[] getIdentifierParts(Identifier identifier) {
-    return StringUtils.split(identifier.getName(), '.');
+    String name = identifier.getName();
+
+    // If identifier represents a top level json object then there must be a '.' in the identifier name.
+    if (name.indexOf('.') != -1) {
```

Comment:
I think we need to check the index of `.` and `[` and pick the one show up the first in order to handle the following cases:
- `json[0].a` = 1
- `json.a[0]` = 1

The output should be:
- `JSON_MATCH(json, '"$[0].a" = 1')`
- `JSON_MATCH(json, '"$.a[0]" = 1')`

We can always split the identifier into 2 parts, and prefix the second part with `$`

Reply:
Fixed. `json[0].a = 1` will map to `JSON_MATCH(json, '"$.[0].a" = 1'`) (note the dot after $). JSON_MATCH / JsonIndex would need to support top-level path expressions as well from what I can see. Added a test case for this.

Reply:
It should be mapped to `JSON_MATCH(json, '"$[0].a" = 1'`. Notice there is no `.` between `$` and `[`. This is standard json path representation

Reply:
Both are standard. The reason for doing $.[0] is that it doesn't require any special handling. For $[0], we would have to explicitly add a check to remove the dot during path generation.